compiler/semantic/expr.go

```diff
@@ -519,64 +519,82 @@ func (a *analyzer) semAssignments(assignments []ast.Assignment, summarize bool)
 	return out, nil
 }
 
-func (a *analyzer) semAssignment(assign ast.Assignment, summarize bool) (dag.Assignment, error) {
+func (a *analyzer) semAssignment(assign ast.Assignment) (dag.Assignment, error) {
 	rhs, err := a.semExpr(assign.RHS)
 	if err != nil {
-		return dag.Assignment{}, fmt.Errorf("rhs of assignment expression: %w", err)
-	}
-	if _, ok := rhs.(*dag.Agg); ok {
-		summarize = true
+		return dag.Assignment{}, fmt.Errorf("right-hand side of assignment expression: %w", err)
 	}
 	var lhs dag.Expr
-	if assign.LHS != nil {
-		lhs, err = a.semExpr(assign.LHS)
+	if assign.LHS == nil {
+		path, err := derriveLHSPath(rhs)
 		if err != nil {
-			return dag.Assignment{}, fmt.Errorf("lhs of assigment expression: %w", err)
+			return dag.Assignment{}, err
+		}
+		lhs = &dag.This{Kind: "This", Path: path}
+	} else {
+		if lhs, err = a.semExpr(assign.LHS); err != nil {
+			return dag.Assignment{}, fmt.Errorf("left-hand side of assignment expression: %w", err)
+		}
+	}
+	if !a.isLval(lhs) {
+		return dag.Assignment{}, errors.New("illegal left-hand side of assignment expression")
+	}
+	if this, ok := lhs.(*dag.This); ok {
+		if len(this.Path) == 0 {
+			return dag.Assignment{}, errors.New("cannot assign to 'this'")
+		}
+	}
+	return dag.Assignment{Kind: "Assignment", LHS: lhs, RHS: rhs}, nil
+}
+
+func (a *analyzer) isLval(e dag.Expr) bool {
+	switch e := e.(type) {
+	case *dag.BinaryExpr:
+		return e.Op == "[" && a.isLval(e.LHS)
+	case *dag.Dot:
+		return a.isLval(e.LHS)
+	case *dag.This:
+		return true
+	}
+	return false
+}
+
+func assignHasDynamicPath(assignments []dag.Assignment) bool {
```

Comment:
Nits: I think this would be a clearer name. And maybe move to op.go since that's the only file in which it's used?
```suggestion
func assignmentHasDynamicLHS(assignments []dag.Assignment) bool {
```